### PR TITLE
Unregister signals for unused views (for better performance)

### PR DESCRIPTION
**Pull request by light-and-salt**
`ved.cql.views` is a `map`, where the `key` is the `id` for each view (e.g., `vis-0-0-0-2-0`) and the `value` is the `view` object.

I noticed that the view `id` is being re-used, which makes it less ideal for `key`. What's a better key for each view? Do views have globally unique names?

### CODE REVIEW

File: app/editor.js [hunk 1]
```diff
@@ -300,6 +301,7 @@ ved.cql.init = function(data) {
   ved.cql.schema = new cql.schema.Schema(fieldSchemas);
   var summary = vg.util.summary(data);
   ved.cql.stats = new cql.stats.Stats(summary);
+  ved.cql.views = new d3.map();
```
Comment (kanitw) [inline app/editor.js]: just use object -- we don't use `d3.map` anywhere else so it's better to be consistent.

File: app/editor.js [hunk 2]
```diff
@@ -445,7 +447,20 @@ ved.cql.renderItems = function(sel, group, indexPrefix) {
...
       };
-      vg.embed(id, opt);
+      vg.embed(id, opt, function(err, result) {
+        if(err) {
+          console.error("[CompassQL]", err);
+          return;
+        }
+
+        // unregister signals for unused views
+        if (ved.cql.views.get(id)) {
```
Comment (kanitw) [inline app/editor.js]: This is wrong.  This only remove views if there are new views of the same id.  
You should call destroy during `exit()`

**Comment by kanitw** [comment]
> I noticed that the view id is being re-used, which makes it less ideal for key. What's a better key for each view? Do views have globally unique names?

Well, the vega-lite spec suppose to be unique.  (We don't enumerate the same view twice.) 
Why don't you use the spec as a key?   (If you loose closely in the code, you'd have seen that [we use stringified specs](https://github.com/vega/vega-editor/blob/compassql/app/editor.js#L323) as key for `.data()` already)

File: app/editor.js (hunk 2 above)
Comment (kanitw) [inline app/editor.js]: One thing that you might wanna do besides removing things in exit() is moving `exit()` to run before `enter()` so removed views already get removed from the dictionary before enter is running here.

File: app/editor.js [hunk 3]
```diff
@@ -314,6 +313,34 @@ function getRankingSummaryText(orderBy, score) {
     }).join(',\n');
 }
 
+/**  
+* Recursively detach event listeners for all views in a group
+* So the event signals can be garbage-collected when a group exits
+*/
+function detachViewsInGroup(root) {
```
Comment (kanitw) [inline app/editor.js]: rename `root` to `item` and rename `item` to `childItem`.

**Comment by kanitw** [comment]
LGTM with minor comment.  Will merge after you address it.

**Comment by light-and-salt** [comment]
Addressed :)

**Pull request merged**